Generated /tmp/nginx-cfg File Invalid When Creating Many Ingress Objects Simultaneously

<!--

Welcome to ingress-nginx!  For a smooth issue process, try to answer the following questions.
Don't worry if they're not all applicable; just try to include what you can :-)

If you need to include code snippets or logs, please put them in fenced code
blocks.  If they're super-long, please use the details tag like
<details><summary>super-long log</summary> lots of stuff </details>

-->

<!--

IMPORTANT!!!

Please complete the next sections or the issue will be closed.
This questions are the first thing we need to know to understand the context.

-->

**NGINX Ingress controller version**: 0.35.0

**Kubernetes version** (use `kubectl version`): v1.18.4

**Environment**: Bare Metal Kubernetes on CentOS 7.6

- **Cloud provider or hardware configuration**: Bare metal Kubernetes
- **OS** (e.g. from /etc/os-release): CentOS 7.6
- **Kernel** (e.g. `uname -a`): 3.10.0-957.1.3.el7.x86_64
- **Install tools**:
- **Others**: Helm v3.1.2

**What happened**:

We have a K8S manifest that includes 100+ Ingress objects. On rare occasions the application of this manifest will fail with an error such as:

`757 2020/08/07 05:41:40 [emerg] 2702#2702: "client_max_body_size" directive is not allowed here in /tmp/nginx-cfg320626064:4
758 nginx: [emerg] "client_max_body_size" directive is not allowed here in /tmp/nginx-cfg320626064:4
759 nginx: configuration file /tmp/nginx-cfg320626064 test failed`

**What you expected to happen**:

The manifest application to succeed, as it does the vast majority of the time.

**How to reproduce it**:
I found issue [5096](https://github.com/kubernetes/ingress-nginx/issues/5096), which includes good reproduction steps. Basically create a helm template directory with two files. The structure and contents should look like:

    # tree helm
    helm
    |-- Chart.yaml
     -- templates
         -- ingress.yaml
    
    1 directory, 2 files
    
    # cat helm/Chart.yaml
    name: testapp1
    version: 0.1.0
    
    # cat helm/templates/ingress.yaml
    {{- range $i, $e := until 50 }}
    apiVersion: extensions/v1beta1
    kind: Ingress
    metadata:
      name: test-ingress-{{$i}}
      labels:
        app: test
        release: test
      annotations:
        kubernetes.io/ingress.class: nginx
        nginx.ingress.kubernetes.io/backend-protocol: "HTTPS"
        nginx.ingress.kubernetes.io/rewrite-target: /$1
        nginx.ingress.kubernetes.io/proxy-body-size: "150m"
        nginx.ingress.kubernetes.io/proxy-connect-timeout: "7200"
        nginx.ingress.kubernetes.io/proxy-read-timeout: "7200"
        nginx.ingress.kubernetes.io/proxy-send-timeout: "7200"
    spec:
      tls:
        - hosts:
          - test.cluster.domain.com
          secretName: ssl
      rules:
      - host: test.cluster.domain.com
        http:
          paths:
          - path: /bp-test/api-{{$i}}
            backend:
              serviceName: test-svc-{{$i}}
              servicePort: 443
    ---
    {{end}}





You can then run:

`while :; do helm upgrade --install testapp helm/; sleep 1; helm delete testapp || true; done`

It should fail pretty frequently; in my case about 25% of the time. 

For one of my failures when it was complaining about directive 'xy_set_header', I grabbed the bad tmp config file from the controller. In it I found:

```
client_max_body_size                    2048m;
xy_set_header X-Forwarded-For           $remote_addr;
proxy_set_header X-Forwarded-Proto      $pass_access_scheme;
```

Obviously the **xy_set_header** is not correct, but this does not appear to be under user control. It seems that ingress-nginx is generating the config file incorrectly when a lot of ingress objects are being created at once.

/kind bug

I too seem to be having the same issue. I am also using Helm. My helm install seems to have tried to create this Ingress service 3 times in quite succession.

The only custom annotation I am using is `nginx.ingress.kubernetes.io/proxy-body-size: 20m`. If I create the resources manually then subsequent Helm upgrades seem to work fine.

Each of them failing:

client.go:108: [debug] creating 1 resource(s)
client.go:108: [debug] creating 7 resource(s)
Release "wintrix-develop" does not exist. Installing it now.
Error: admission webhook "validate.nginx.ingress.kubernetes.io" denied the request: 
-------------------------------------------------------------------------------
Error: exit status 1
2021/01/11 11:56:19 [emerg] 359#359: "proxy_set_header" directive is not allowed here in /tmp/nginx-cfg330571149:2795
nginx: [emerg] "proxy_set_header" directive is not allowed here in /tmp/nginx-cfg330571149:2795
nginx: configuration file /tmp/nginx-cfg330571149 test failed

-------------------------------------------------------------------------------

helm.go:84: [debug] admission webhook "validate.nginx.ingress.kubernetes.io" denied the request: 
-------------------------------------------------------------------------------
Error: exit status 1
2021/01/11 11:56:19 [emerg] 359#359: "proxy_set_header" directive is not allowed here in /tmp/nginx-cfg330571149:2795
nginx: [emerg] "proxy_set_header" directive is not allowed here in /tmp/nginx-cfg330571149:2795
nginx: configuration file /tmp/nginx-cfg330571149 test failed

-------------------------------------------------------------------------------

##[error]history.go:52: [debug] getting history for release wintrix-develop
install.go:159: [debug] Original chart version: "10603"
install.go:176: [debug] CHART PATH: /home/buildadmin/.cache/helm/repository/wintrix-10603.tgz

client.go:108: [debug] creating 1 resource(s)
client.go:108: [debug] creating 7 resource(s)
Error: admission webhook "validate.nginx.ingress.kubernetes.io" denied the request: 
-------------------------------------------------------------------------------
Error: exit status 1
2021/01/11 11:56:19 [emerg] 359#359: "proxy_set_header" directive is not allowed here in /tmp/nginx-cfg330571149:2795
nginx: [emerg] "proxy_set_header" directive is not allowed here in /tmp/nginx-cfg330571149:2795
nginx: configuration file /tmp/nginx-cfg330571149 test failed

-------------------------------------------------------------------------------

helm.go:84: [debug] admission webhook "validate.nginx.ingress.kubernetes.io" denied the request: 
-------------------------------------------------------------------------------
Error: exit status 1
2021/01/11 11:56:19 [emerg] 359#359: "proxy_set_header" directive is not allowed here in /tmp/nginx-cfg330571149:2795
nginx: [emerg] "proxy_set_header" directive is not allowed here in /tmp/nginx-cfg330571149:2795
nginx: configuration file /tmp/nginx-cfg330571149 test failed

-------------------------------------------------------------------------------

--- a/internal/ingress/controller/template/template.go
+++ b/internal/ingress/controller/template/template.go
@@ -63,6 +63,9 @@ const (
 
 // Writer is the interface to render a template
 type Writer interface {
+	// Write renders the template.
+	// NOTE: Implementors must ensure that the content of the returned slice is not modified by the implementation
+	// after the return of this function.
 	Write(conf config.TemplateConfig) ([]byte, error)
 }
 
@@ -202,7 +205,12 @@ func (t *Template) Write(conf config.TemplateConfig) ([]byte, error) {
 		return nil, err
 	}
 
-	return outCmdBuf.Bytes(), nil
+	// make a copy to ensure that we are no longer modifying the content of the buffer
+	out := outCmdBuf.Bytes()
+	res := make([]byte, len(out))
+	copy(res, out)
+
+	return res, nil
 }
 
 var (

